### ISSUE TITLE
Remove Dish Stirling & Direct Steam Power tower from Startup Script

### ISSUE THREAD

**Issue by janinefreeman** [NREL/SAM#363]
The dish stirling and direct steam technologies in SAM have become outdated and do not have current funding support to continue updating, so we are planning to remove them from the SAM user interface for our fall 2020 release.

**Comment by dguittet** [comment]
@mjprilliman @sjanzou How to handle blank case with no tabs? Or an old version SAM project file that doesn't have these removed?

**Comment by cpaulgilman** [comment]
@mjprilliman, What branch did you use for your work on this?

**Comment by mjprilliman** [comment]
remove_csp_technologies. 

I have it checked out right now. Looking further I am seeing some issues pop up where metrics.lk, autographs.lk, and cashflow.lk are looking for the financial model information contained in the retired technologies when opening projects from an older version. So we either need to go back to figuring out how to comment out the technologies in startup without SAM crashing or I am looking at how to delete the case as part of version upgrade but so far no success.

**Comment by cpaulgilman** [comment]
These possible solutions for opening an old SAM case that contains a retired technology would require writing new functions for the Version Upgrade flavor of LK:

1. Display a `msgbox()` saying that this version of SAM cannot open the file because it contains a case with a retired model and then quit SAM. (Include text in message box to explain that if the file contains other cases with non-retired models, the user can use File->Import to open those cases.) Requires a new `quit_sam()` function.

2. Remove any cases with retired models from the file and notify the user using `message()`. If the file only contains a single case with a retired model, close SAM so that the Welcome page is displayed. Requires new `delete_case()' and `close_sam()` functions.

These approaches would mean that every time we retire a case, we have to modify the version upgrade script to handle the retired models, so more effort than simply commenting out lines in the startup script. Also, for each new version of SAM, we would have to be sure to include the code that checks for retired models in the function for that version.

**Comment by mjprilliman** [comment]
@cpaulgilman Take a look at my latest push on removing_csp_technologies. I couldn't figure out how to make it into an LK function yet but I have a start on option 2. My only concern would be people losing their case data if they save the upgraded project file, already there is already a two-step prompt when saving in this situation.

**Comment by mjprilliman** [comment]
@sjanzou @cpaulgilman The current solution is workable but requires adjustments to the source code for future model retirements. I haven't yet figured out a way to achieve the same behavior by changing the LK scripts so any ideas would be appreciated.

**Comment by mjprilliman** [comment]
@cpaulgilman @sjanzou My new idea on this issue is to write a lk function in the version upgrade flavor that rewrites retired technologies 'technology' field to a dummy technology known as 'Retired'. Right now this check is happening based on the technology string from the case, but it could potentially change to check for the tree_parent in startup.lk, which I am also changing to 'Retired' for the select technologies for populateTech checking and menu building. 

I still have some code cleanup and I still need to figure out how to close SAM when an upgrade contains only retired technologies, but is this a desirable path forward?

![image](https://user-images.githubusercontent.com/54449384/98021713-df77dc00-1dc1-11eb-8152-94059eb7255a.png)

![image](https://user-images.githubusercontent.com/54449384/98021825-059d7c00-1dc2-11eb-930a-346a03c7dd5f.png)

![image](https://user-images.githubusercontent.com/54449384/98021864-13eb9800-1dc2-11eb-8152-2a4a0b97be29.png)

**Comment by mjprilliman** [comment]
There is still behavior in the upgrade script where if a file that is being upgrade to the latest version only contains retired technologies, the case is deleted but the SAM UI is left on a blank screen. I have not yet found a way to force my way back to the welcome screen from this point. Users can still open a new file or add a case to their existing file that contains no cases currently.

**Comment by sjanzou** [comment]
> There is still behavior in the upgrade script where if a file that is being upgrade to the latest version only contains retired technologies, the case is deleted but the SAM UI is left on a blank screen. I have not yet found a way to force my way back to the welcome screen from this point. Users can still open a new file or add a case to their existing file that contains no cases currently.

@mjprilliman, the blank project file not returning to the welcome screen is the same issue that occurs when the last case in a project file is deleted. I will work on fixing both issues. 

I checked in a change to the remove_csp_technologies branch that was throwing an exception in debug mode when trying to call RecalculateAll on a deleted case.

I will review your methodology more and let you know if I find any issues!

**Comment by sjanzou** [comment]
Project file to test and current status:

[DS_Com.zip](https://github.com/NREL/SAM/files/5487995/DS_Com.zip)

[GitHubIssuesSAM_363.docx](https://github.com/NREL/SAM/files/5487997/GitHubIssuesSAM_363.docx)

**Comment by sjanzou** [comment]
@mjprilliman, @cpaulgilman, This issue should be ready to be closed with pull request #387. Please do not merge files other than project.cpp, main.cpp and versions.lk into develop. Tested per attached document with prompt to handle project files with no cases

[GitHubIssuesSAM_363.docx](https://github.com/NREL/SAM/files/5494393/GitHubIssuesSAM_363.docx)

**Comment by cpaulgilman** [comment]
Use cases:

1. If a SAM file has one case, and you try to delete the case, SAM should not close the case and instead display a message like "Cannot delete case when there is only one case in the file." SAM should not allow "empty" files with no cases.

2. If you open a file that contains only retired cases, SAM should display a message like "Cannot open a file that contains only retired cases." and then either show the Welcome page (ideal) or quit without overwriting the file.

**Comment by mjprilliman** [comment]
[2retiredtechs_1pvtech.zip](https://github.com/NREL/SAM/files/5498024/2retiredtechs_1pvtech.zip)

Case file with a dish stirling case (retired), direct steam power tower case (retired), and pv case saved in 2018 version.

**Comment by cpaulgilman** [comment]
In versions.lk, I suggest checking for retired models first and returning rather than continuing the upgrade process. This makes the upgrade message cleaner.

![image](https://user-images.githubusercontent.com/27711933/98450164-48779080-20ef-11eb-863a-da7a9dac21b4.png)


```
version_upgrade{'2020.11.3'} = define() {

    const cfg = config();

    //To be included in next full release (version_upgrade{'2020.10.?'})
    if (cfg.tech == 'Dish Stirling' )
    {
		retire_tech();
		//The version upgrade script is set up to delete cases with this technology made from a previous release from the project file. This is done in project.cpp
		message('The Dish Stirling model is not available in SAM ' + samver() +'. Removing Case \"' + case_name() + '\" from the file. The most recent version with the model is SAM 2020.2.29, which you can download from https://sam.nrel.gov/download/','warning');
		return;
    }
    if (cfg.tech == 'DSPT')
    {
		retire_tech();
		//The version upgrade script is set up to delete cases with this technology made from a previous release from the project file. This is done in project.cpp
		message ('The Direct Steam Power Tower model is not available in SAM ' + samver() +'. Removing Case \"' + case_name() + '\" from the file. The most recent version with the model is SAM 2020.2.29, which you can download from https://sam.nrel.gov/download/','warning');
		return;
	}

...

```

**Comment by sjanzou** [comment]
@mjprilliman, @cpaulgilman please review PR #387. Should close this issue.
